crates/re_query2/src/latest_at/results.rs

```diff
@@ -0,0 +1,221 @@
+use nohash_hasher::IntMap;
+use re_log_types::{DataCell, RowId, TimeInt};
+use re_types_core::ComponentName;
+use re_types_core::{Component, DeserializationError, DeserializationResult};
+
+use crate::{Promise, PromiseResolver, PromiseResult};
+
+// ---
+
+/// Raw results for a latest-at query.
+///
+/// The data is neither deserialized, nor resolved/converted.
+/// It it the raw [`DataCell`]s, straight from our datastore.
+///
+/// Use [`LatestAtResults::get`], [`LatestAtResults::get_required`] and [`LatestAtResults::get_optional`]
+/// in order to access the raw results for each individual component.
+#[derive(Debug, Clone)]
+pub struct LatestAtResults {
+    /// The compound index of this query result.
+    ///
+    /// A latest-at query is a compound operation that gathers data from many different rows.
+    /// The index of that compound result corresponds to the index of most the recent row in all the
+    /// sub-results, as defined by time and row-id order.
+    pub compound_index: (TimeInt, RowId),
+
+    /// Raw results for each individual component.
+    pub components: IntMap<ComponentName, LatestAtComponentResults>,
+}
+
+impl Default for LatestAtResults {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            compound_index: (TimeInt::STATIC, RowId::ZERO),
+            components: Default::default(),
+        }
+    }
+}
+
+impl LatestAtResults {
+    #[inline]
+    pub fn contains(&self, component_name: impl Into<ComponentName>) -> bool {
+        self.components.contains_key(&component_name.into())
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    #[inline]
+    pub fn get<C: Component>(&self) -> Option<&LatestAtComponentResults> {
+        self.components.get(&C::name())
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    ///
+    /// Returns an error if the component is not present.
+    #[inline]
+    pub fn get_required<C: Component>(&self) -> crate::Result<&LatestAtComponentResults> {
+        if let Some(component) = self.components.get(&C::name()) {
+            Ok(component)
+        } else {
+            Err(DeserializationError::MissingComponent {
+                component: C::name(),
+                backtrace: ::backtrace::Backtrace::new_unresolved(),
+            }
+            .into())
+        }
+    }
+
+    /// Returns the [`LatestAtComponentResults`] for the specified [`Component`].
+    ///
+    /// Returns empty results if the component is not present.
+    #[inline]
+    pub fn get_optional<C: Component>(&self) -> &LatestAtComponentResults {
```

Comment:
The naming of this vs `get` is a bit confusing at first. `get` follows the semantics of other Rust containers (i.e. returning an `Option`), so that makes sense. However, I would instinctually think something called `get_optional` would also return an `Option`, but it doesn't.

EDIT: but it make sense from how its used, since this is the de-factor _normal_ way of accessing an optional component

Reply:
I might go for something like `get_or_empty` ye